### PR TITLE
[CUDA] work around more __noinline__ conflicts with libc++

### PR DESCRIPTION
**Pull request by Artem-B**
https://github.com/llvm/llvm-project/pull/73838

### CODE REVIEW

**Comment by gribozavr** [comment]
I don't mind this, but should libc++ start using `_LIBCPP_NOINLINE`, we would be playing a catch-up game here.

**Comment by Artem-B** [comment]
Yes, I've mentioned that in https://github.com/llvm/llvm-project/pull/73838. However, we need something to fix the issue right now while we're figuring out a better solution. 

In any case `__noinline__` is unlikely to be widely used, so the wrappers may be manageable, at least for the short-to-medium term.

File: clang/lib/Headers/cuda_wrappers/string [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// CUDA headers define __noinline__ which interferes with libc++'s use of
+// `__attribute((__noinline__))`. In order to avoid compilation error,
+// temporarily unset __noinline__ when we include affected libstdc++ header.
```
Comment (philnik777) [inline clang/lib/Headers/cuda_wrappers/string:3]:
```suggestion
// temporarily unset __noinline__ when we include affected libc++ header.
```

**Review by philnik777 — COMMENTED** [review]
I think we can find a solution to work around this in libc++ within a reasonable timeframe, but I won't block this, since I don't really care about this part of the code base.

**Comment by Artem-B** [comment]
> I think we can find a solution to work around this in libc++ within a reasonable timeframe

OK. I'll hold off on landing the patch. I believe we're not blocked on it at the moment.

**Review by ldionne — COMMENTED** [review]
Some review comments if we end up merging this.

FWIW I am not thrilled about using `__config` here. That header is an implementation detail of libc++ and defining it and relying on it is somewhat brittle.

**Comment by Artem-B** [comment]
> FWIW I am not thrilled about using `__config` here. That header is an implementation detail of libc++ and defining it and relying on it is somewhat brittle.

I'm all for having it fixed in libc++ or in CUDA SDK. Barring that, working around the specific implementation details is all I can do here.  I'm open to other ideas on how this conflict can be worked around or solved in a better way.

**Comment by gribozavr** [comment]
https://github.com/llvm/llvm-project/pull/73838 has landed, we can close this PR now.

**Pull request closed without merging**